File: Suri/rei2/rei2-18/rei2-18.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.cm as cm

#それぞれのテキストファイルからデータを読み込む
data0 = np.genfromtxt("data0.txt")
data1 = np.genfromtxt("data1.txt")
data2 = np.genfromtxt("data2.txt")
data3 = np.genfromtxt("data3.txt")
data4 = np.genfromtxt("data4.txt")
data5 = np.genfromtxt("data5.txt")
data6 = np.genfromtxt("data6.txt")
data7 = np.genfromtxt("data7.txt")


#画像を出力するための領域を作成
fig = plt.figure()

#領域の中でプロットする場所を決める
ax = fig.add_subplot(1,1,1)

#領域の中でプロットする場所を決める
ax.plot(data0[:,0],data0[:,1],color=cm.hsv(1/8.0),label="t=0.000")
ax.plot(data1[:,0],data1[:,1],color=cm.hsv(2/8.0),label="t=0.001")
ax.plot(data2[:,0],data2[:,1],color=cm.hsv(3/8.0),label="t=0.002")
ax.plot(data3[:,0],data3[:,1],color=cm.hsv(4/8.0),label="t=0.004")
ax.plot(data4[:,0],data4[:,1],color=cm.hsv(5/8.0),label="t=0.008")
ax.plot(data5[:,0],data5[:,1],color=cm.hsv(6/8.0),label="t=0.016")
ax.plot(data6[:,0],data6[:,1],color=cm.hsv(7/8.0),label="t=0.032")
ax.plot(data7[:,0],data7[:,1],color=cm.hsv(8/8.0),label="t=0.064")


# x,y軸に補助目盛線を設定
ax.grid(which = "major", axis = "both", alpha = 0.8,linestyle = "--", linewidth = 1)

# x,y軸の目盛設定
ax.set_xlabel("x",weight="bold",size="xx-large")
ax.set_ylabel("y",weight="bold",size="xx-large")

ax.set_xticks(np.arange(0,1.2,0.25))
#凡例を追加
ax.legend()

#プロットしたグラフを画像形式に保存する
plt.savefig("rei2-18.png")

#プロットしたグラフを出力する
plt.show()

File: Suri/rei2/rei2-18/rei2-18.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>

int FNU(double, double, double);
int FNV(double, double, double);

int main(){
    double dx = 0.01;
    double t = 0;
    double dt = 0.00001;
    double Y[101];
    int n = 0;
    FILE *fpout;

    for(int i = 0;i <= 100;i++){
        if(i >= 25 && i <= 50){
            Y[i] = 12;
        }else if(i >= 75 && i <= 90){
            Y[i] = 7;
        }else{
            Y[i] = 0;
        }
    }
    

    n = 6400;
    for(int k = 1; k <= n; k++){
        if(t == 0){
            fpout = fopen("data0.txt","w");
            fprintf(fpout,"# x = , y = ,(t = 0.000)\n");
            for(int l = 1;l <= 100;l++){
                fprintf(fpout,"%lf %lf\n",l*dx,Y[l]);
            }
        }
        for(int i = 1; i <= 99; i++){
            Y[i] = Y[i] + ((Y[i+1] - 2*Y[i] + Y[i-1])*dt)/pow(dx,2);
        }
        Y[0]=Y[1];
        Y[100]=Y[99];
        switch(k){
            case 100:
                fpout = fopen("data1.txt","w");
                fprintf(fpout,"# x = , y = ,(t = 0.001)\n");
                for(int l = 1;l <= 100;l++){
                    fprintf(fpout,"%lf %lf\n",l*dx,Y[l]);
                }
                t = t + dt;
                break;

            case 200:
                fpout = fopen("data2.txt","w");
                fprintf(fpout,"# x = , y = ,(t = 0.002)\n");
                for(int l = 1;l <= 100;l++){
                    fprintf(fpout,"%lf %lf\n",l*dx,Y[l]);
                }
                t = t + dt;
                break;

            case 400:
                fpout = fopen("data3.txt","w");
                fprintf(fpout,"# x = , y = ,(t = 0.004)\n");
                for(int l = 1;l <= 100;l++){
                    fprintf(fpout,"%lf %lf\n",l*dx,Y[l]);
                }
                t = t + dt;
                break;

            case 800:
                fpout = fopen("data4.txt","w");
                fprintf(fpout,"# x = , y = ,(t = 0.008)\n");
                for(int l = 1;l <= 100;l++){
                    fprintf(fpout,"%lf %lf\n",l*dx,Y[l]);
                }
                t = t + dt;
                break;

            case 1600:
                fpout = fopen("data5.txt","w");
                fprintf(fpout,"# x = , y = ,(t = 0.016)\n");
                for(int l = 1;l <= 100;l++){
                    fprintf(fpout,"%lf %lf\n",l*dx,Y[l]);
                }
                t = t + dt;
                break;

            case 3200:
                fpout = fopen("data6.txt","w");
                fprintf(fpout,"# x = , y = ,(t = 0.032)\n");
                for(int l = 1;l <= 100;l++){
                    fprintf(fpout,"%lf %lf\n",l*dx,Y[l]);
                }
                t = t + dt;
                break;

            case 6400:
                fpout = fopen("data7.txt","w");
                fprintf(fpout,"# x = , y = ,(t = 0.064)\n");
                for(int l = 1;l <= 100;l++){
                    fprintf(fpout,"%lf %lf\n",l*dx,Y[l]);
                }
                t = t + dt;
                break;

            case 12800:
                fpout = fopen("data8.txt","w");
                fprintf(fpout,"# x = , y = ,(t = 0.064)\n");
                for(int l = 1;l <= 100;l++){
                    fprintf(fpout,"%lf %lf\n",l*dx,Y[l]);
                }
                t = t + dt;
                break;

            case 25600:
                fpout = fopen("data9.txt","w");
                fprintf(fpout,"# x = , y = ,(t = 0.064)\n");
                for(int l = 1;l <= 100;l++){
                    fprintf(fpout,"%lf %lf\n",l*dx,Y[l]);
                }
                t = t + dt;
                break;

            default:
                t = t + dt;
        }
        
    }
    fclose(fpout);
    return 0;
}
